lib/ansible/modules/cloud/amazon/emr_cluster.py

```diff
@@ -0,0 +1,363 @@
+      type: list
+    tags:
+      description: list of tags
+      type: dict
+    service_role:
+      description: IAM role that will be assumed by the Amazon AEMR service
+      type: str
+    master_public_dns_name:
+      description: the dns name of the master node
+      type: str
+    configurations:
+      description: the list of configurations supplied to EMR cluster
+      type: dict
+'''
+
+try:
+    from botocore.exception import ClientError
+except ImportError:
+    pass  # will be picked up from imported HAS_BOTO3
+
+import re
+import traceback
+import time
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.ec2 import ec2_argument_spec, boto3_conn, get_aws_connection_info
+from ansible.module_utils.ec2 import HAS_BOTO3, camel_dict_to_snake_dict, snake_dict_to_camel_dict, ansible_dict_to_boto3_tag_list
+
+
+def terminate_emr_cluster(client, module):
+    cluster_id = module.params.get('cluster_id')
+    wait = module.params.get('wait')
+    wait_timeout = module.params.get('wait_timeout')
+
+    if not cluster_id:
+        module.fail_json(msg='cluster_id required for state absent')
+
+    client.terminate_job_flows(JobFlowIds=[cluster_id])
+
+    resource = client.describe_cluster(ClusterId=cluster_id)['Cluster']
+
+    if wait:
+        wait_timeout = time.time() + wait_timeout
+        time.sleep(5)
+
+        while wait_timeout > time.time() and resource['Status']['State'] != 'TERMINATED':
+            time.sleep(5)
+
+            if wait_timeout <= time.time():
+                module.exit_json(msg="Timeout waiting for resource %s" % resource['Id'])
+
+            resource = client.describe_cluster(ClusterId=cluster_id)['Cluster']
+
+    return camel_dict_to_snake_dict(resource)
+
+
+def create_emr_cluster(client, module):
+    params = {}
+    wait = module.params.get('wait')
+    wait_timeout = module.params.get('wait_timeout')
+
+    for p in ['name', 'instances', 'log_uri', 'additional_info', 'ami_version', 'release_label',
+              'steps', 'bootstrap_actions', 'supported_products', 'new_supported_products',
+              'applications', 'configurations', 'visible_to_all_users', 'service_role', 'job_flow_role',
+              'security_configuration', 'auto_scaling_role', 'scale_down_behavior', 'custom_ami_id',
+              'repo_upgrade_on_boot', 'ebs_root_volume_size', 'cluster_id', 'kerberos_attributes']:
+
+        if p in module.params and module.params.get(p):
+            params[p] = module.params.get(p)
+
+    if 'tags' in module.params and module.params.get('tags'):
+        params['tags'] = ansible_dict_to_boto3_tag_list(module.params.get('tags'))
+
+    if not params['name']:
+        module.fail_json(msg="name required for the state create")
+
+    params = snake_dict_to_camel_dict(params, capitalize_first=True)
+
+    job_flow_id = client.run_job_flow(**params)['JobFlowId']
+
+    resource = client.describe_cluster(ClusterId=job_flow_id)['Cluster']
+
+    if wait:
```

Comment:
There is a waiter for this too: http://boto3.readthedocs.io/en/latest/reference/services/emr.html#EMR.Waiter.ClusterRunning